rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -662,3 +812,178 @@ def prediction_states(
+        The number of columns is equal to the maximum number of labels
+        that any Labels item has. Rows are padded with -1 if not all Labels
+        items have the same number of labels.
+
+        Returns:
+            A 2d np.ndarray of label ids.
+        """
+
+        # store labels in numpy arrays so that it corresponds to np arrays
+        # of input features
+        label_ids = [
+            [domain.intents.index(intent) for intent in tracker_intents]
+            for tracker_intents in trackers_as_intents
+        ]
+
+        pad_val = -1
+
+        # Add -1 padding to labels array so that
+        # each example has equal number of labels
+        multiple_labels_count = [len(a) for a in label_ids]
+        max_labels_count = max(multiple_labels_count)
+        num_padding_needed = [max_labels_count - len(a) for a in label_ids]
+
+        new_label_ids = []
+        for ids, num_pads in zip(label_ids, num_padding_needed):
+            if num_pads:
+                ids.extend([pad_val] * num_pads)
+            new_label_ids.append(ids)
+
+        new_label_ids = np.array(new_label_ids)
+        return new_label_ids
+
+    def _setup_example_iterator(self) -> None:
+        """Create any data structures for deduplication and tracking multiple
+        intent labels.
+        """
+        super()._setup_example_iterator()
+        self._state_hash_to_labels = defaultdict(list)
+
+    def _example_iterator(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        omit_unset_slots: bool = False,
+    ) -> Iterator[Tuple[List[State], List[Text], List[Dict[Text, Any]]]]:
+        """Create an iterator over the training examples that will be created
+        from the provided tracker.
+
+        Returns:
+            An iterator over state, labels, entity tag tuples
+        """
+
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+
+        label_index = 0
+        for event in tracker.applied_events():
+
+            if isinstance(event, ActionExecuted):
+                label_index += 1
+
+            elif isinstance(event, UserUttered):
+
+                sliced_states = self.slice_state_history(
+                    tracker_states[:label_index], self.max_history
+                )
+                label = [event.intent_name or event.intent_text]
+                entities = [{}]
+
+                yield sliced_states, label, entities
+
+    def _check_example_cache(
+        self, tracker: DialogueStateTracker, states: List[State], label: List[Text],
+    ) -> bool:
+        if not super()._check_example_cache(tracker, states, label):
+            state_hash = self._hash_example(tracker, states)
+            self._state_hash_to_labels[state_hash].append(label)
+            return False
+        else:
+            return True
+
+    def _cleanup_example_iterator(self) -> None:
```

Comment:
this is really really difficult to wrap ones head around - even just thinking about the types of the variables and what they might represent is challenging.

I keep repeating myself, but I don't like the side effects. It's also not self evident that a cleanup method actually changes the extracted labels. I would challenge you to get this seemingly important logic out of the cleanup into the example extraction procedures.